Convenient interface types for the Nbody gravitational problem

Mikhail-Vaganov opened the pull request:
It is convenient to create `MassBody` and `NBodyGravProblem` to be simulated instead of mere arrays. Nevertheless, in order to be solved an `NBodyGravProblem` object should be converted to an `ODEProblem`.

Mikhail-Vaganov committed:
Structs for a solution of the gravitational n-body problem

Created two structs for bodies and the problem itself. Also added two visualizing functions.

Mikhail-Vaganov committed:
Modified readme file

Mikhail-Vaganov committed:
Update README.md

Mikhail-Vaganov committed:
Small refactoring of code and the notebook

Mikhail-Vaganov committed:
Tried to change the size of the first image

Mikhail-Vaganov committed:
Update README.md

Smaller images

Mikhail-Vaganov committed:
Added kwargs to the solve function for NBodyGravProblem

Mikhail-Vaganov committed:
Merge branch 'gsoc-nbody-problem' of https://github.com/Mikhail-Vaganov/DiffEqPhysics.jl into gsoc-nbody-problem

Mikhail-Vaganov committed:
Optimized the ode_system function

Changed usage of vectorized data: use SVectors, view macros, etc.
Added a gif for the 5-body choreography taken from Carles Simo's article of 2000.
Updated readme and notebook.

Mikhail-Vaganov committed:
Removed the figure 8 gif, removed padding for the 5-body gif

Mikhail-Vaganov committed:
Update README.md

Mikhail-Vaganov committed:
Removed unnecessary files

Mikhail-Vaganov committed:
Merge branch 'gsoc-nbody-problem' of https://github.com/Mikhail-Vaganov/DiffEqPhysics.jl into gsoc-nbody-problem

ChrisRackauckas commented on nbody_gravitational/NBodyGravitational.jl:
```diff
@@ -0,0 +1,173 @@
+using DiffEqBase, DifferentialEquations, StaticArrays, Plots, StaticArrays
+
+
+grav_const = 6.673e-11
```
Are there units where this is more numerically okay? This kind of a constant may cause cancellation

Mikhail-Vaganov replied:
Sure, but this one is in SI units as a default value. We can set the desired value in constructor of NBodyGravProblem. If  I change artificially `grav_const` to a greater value during computation, the multiplication of stellar masses with this constant will give rather big number - it may also cause problems.

Mikhail-Vaganov replied:
I removed this const - it is unnecessary here (see further)

ChrisRackauckas commented on nbody_gravitational/NBodyGravitational.jl:
```diff
@@ -0,0 +1,173 @@
+=#
+struct MassBody{mType, cType}
+    m :: mType
+    r :: Vector{cType}
```
What about using static arrays to make this a bits value type?

Mikhail-Vaganov replied:
Yes, good question. I used to use `SVector` for these data (`r` and `v`), but if there are a lot of bodies, they take quite big place in the stack memory. Say, 1000 bodies and 1000x2x3 elements to store. Also, `MassBody` is just an intermediate container and doesn't really play role in actual calculation of the differential problem. That is why I use mere `Vector`s.

ChrisRackauckas replied:
A `Vector{SMatrix}` of length 1000 just has the same memory layout as a 3D tensor. It wouldn't have issues with stack memory. You'd only have issues here if it had hundreds of spatial dimensions.

Mikhail-Vaganov replied:
Ok I changed to SVectors -  solution is evaluated quicker.

ChrisRackauckas commented on nbody_gravitational/NBodyGravitational.jl:
```diff
@@ -0,0 +1,173 @@
+    v :: Vector{cType}
+    
+    function MassBody{mType,cType}(m::mType, r::Vector{cType}, v::Vector{cType})
+        @assert length(r)==length(v) "The length of the position (r) and velocity (v) components should match"
```
If you do a static array then this can be type checked for free

Mikhail-Vaganov replied:
Ok

ChrisRackauckas commented on nbody_gravitational/NBodyGravitational.jl:
```diff
@@ -0,0 +1,173 @@
+    end
+end
+
+MassBody(m::mType, r :: Vector{cType}, v :: Vector{cType}) where {mType <: AbstractFloat, cType  <: AbstractFloat} = MassBody{mType, cType}(m,r,v)
```
I don't think this is needed

Mikhail-Vaganov replied:
I agree. Removed.

ChrisRackauckas commented on nbody_gravitational/NBodyGravitational.jl:
```diff
@@ -0,0 +1,173 @@
+=#
+struct NBodyGravProblem
+    bodies :: Array{MassBody, 1}
+    tSpan :: Tuple{Float64, Float64}
```
Lower case. Julia doesn't camelcase in it's standard style (unless it's a module)

Mikhail-Vaganov replied:
Got it.

ChrisRackauckas commented on nbody_gravitational/NBodyGravitational.jl:
```diff
@@ -0,0 +1,173 @@
+struct NBodyGravProblem
+    bodies :: Array{MassBody, 1}
+    tSpan :: Tuple{Float64, Float64}
+    G :: Float64
```
Is there a reason this is localized? While normally globals are bad, here a global constant makes sense.

Mikhail-Vaganov replied:
I wanted to create `NBodyGravProblem` as an object representing the task: with its own defined bodies, constants and units of measurements. That is why I let a user define this gravitational constant via constructor of `NBodyGravProblem`.

ChrisRackauckas commented on nbody_gravitational/NBodyGravitational.jl:
```diff
@@ -0,0 +1,173 @@
+    following by components of velocities v
+=#
+import Base.convert
+convert(::Type{ODEProblem}, x:: NBodyGravProblem) =
```
Use the standard multiline function syntax here.

ChrisRackauckas replied:
Instead of defining convert, define an outer constructor for ODEProblem. Convert falls back to the constructor but not the other way around.

Mikhail-Vaganov replied:
Ok.

ChrisRackauckas commented on nbody_gravitational/NBodyGravitational.jl:
```diff
@@ -0,0 +1,173 @@
+    
+    for i=1:n
+        ith = 3(i-1)+1;        
+        u0[ith:ith+2] = x.bodies[i].r;
```
Maybe it's easier to make this a matrix?

Mikhail-Vaganov replied:
It is truly so! Done.

ChrisRackauckas commented on nbody_gravitational/NBodyGravitational.jl:
```diff
@@ -0,0 +1,173 @@
+            ri = @SVector [u[ith] , u[ith+1], u[ith+2]]
+            du[3n+ith:3n+ith+2] = 
+            begin
+                accel = @SVector [0.0, 0.0, 0.0]; 
```
The syntax is a little odd. Just calculate accel and then write the to du?

Mikhail-Vaganov replied:
Changed it.

ChrisRackauckas commented on nbody_gravitational/NBodyGravitational.jl:
```diff
@@ -0,0 +1,173 @@
+end
+
+import DiffEqBase.solve
+solve(x::NBodyGravProblem; kwargs...) = 
```
Use standard function syntax here

ChrisRackauckas replied:
Optionally allow an algorithm choice

Mikhail-Vaganov replied:
Done.

ChrisRackauckas commented on nbody_gravitational/NBodyGravitational.jl:
```diff
@@ -0,0 +1,173 @@
+import DiffEqBase.solve
+solve(x::NBodyGravProblem; kwargs...) = 
+begin
+    return DiffEqBase.solve(convert(ODEProblem,x); alg_hints=[:stiff], kwargs...)
```
Is it usually stiff?

Mikhail-Vaganov replied:
I suppose so: if objects move quite close to each other, the step of integration should decrease dramatically. It is a general problem of potentials with singularity in denominator (as Coulomb's one). Plus, real-life number of masses and gravitational constant can also bring errors.  So, by default, I reckon this problem is stiff.

ChrisRackauckas replied:
I would test it before making that assumption. N-body problems like the Pleiades problem are classic non-stiff tests for ODE integrators.

http://nbviewer.jupyter.org/github/JuliaDiffEq/DiffEqBenchmarks.jl/blob/master/NonStiffODE/Pleiades%20Work-Precision%20Diagrams.ipynb

Also, many symplectic RK integrators were created for these problems, and they are for non-stiff problems. I bet you can find examples of stiff N-body problems, but I would think it's rare given the benchmarks people use. Of course, this is a good reason just just splat the arguments.

ChrisRackauckas commented on nbody_gravitational/NBodyGravitational.jl:
```diff
@@ -0,0 +1,173 @@
+=#
+import Base.convert
+convert(::Type{ODEProblem}, x:: NBodyGravProblem) =
+begin
```
It probably makes sense to define the constructor for a SecondOrderODEProblem

Mikhail-Vaganov replied:
I agree. I have added it.

ChrisRackauckas commented on nbody_gravitational/NBodyGravitational.jl:
```diff
@@ -0,0 +1,173 @@
+end
+
+# Create a gif file representing moving bodies and their orbits
+function plot_xy_scattering(solution::ODESolution, path::AbstractString, duration::AbstractFloat = 3.0; plot_kwargs...)
```
These should be recipes to not add the dependency. Thats a whole deal though

ChrisRackauckas reviewed (changes requested):
One big comment: do we need to implement each potential function differently? What do we gain by doing so?

ChrisRackauckas commented:
It's looking great! I left some comments to guide it along. The main thing I would like to discuss is what about this approach is specialized specifically for gravitational problems, and whether this can be structured to handle more general potential functions. Likely there's a lot that can be done but we should continue that in a future PR, but it would be good to start planning it out.

Mikhail-Vaganov committed:
Refactoring after review

- Returned to Static Arrays
- Added tests
- Added constructor for `SecondOrderODEProblem`
- Use table for `u` in ODEProblem istead of vector

Mikhail-Vaganov commented:
1. I tried to add a test for my 'NBodyGravProblem' - probably I have done something wrong, because tests have not passed. I definitely need help in writing tests in Julia. Is there any tutorial? 
2. Regarding specialization of this code specifically for gravitational problem: it so, there is no doubt. Here are my thoughts upon possible solutions for generalising the tools:
    1. Develop an interface allowing one to create bodies/particles in convenient manner and to define a potential function accepting those particles as one of its arguments. This potential function will then be used in a constructor for an `NBodyProblem(particles, potential_func, u0, v0)` - something like that.
    2. Develop a bunch of physically valid potentials and switch them on for particular object of `NBodyProblem`. Also provide a particle-struct with fields for common physical quantities like charge, mass, magnetic moment, etc.  For example, if we want to allow for electrostatic interaction during simulations, we could create object  `NBodyProblem(particles, u0, v0, potentials=[:electrostatic])`.
   3. Though standard potentials might be the most common and widely used, often it is necessary to change even a standard potential slightly. Therefore, we might consider combining both approaches above.
   4. I think it is also possible sometimes to operate in terms of forces. For example, in `NBodyGravProblem` one can add another acceleration component in addition to `gravitational_acceleration`, which in its turn might be switched on or off via some parameter/argument.

Mikhail-Vaganov commented:
Regarding tests: I found, that there was an error 

> ERROR: LoadError: LoadError: LoadError: ArgumentError: Module DifferentialEquations not found in current path.
Run `Pkg.add("DifferentialEquations")` to install the DifferentialEquations package.

What should be the correct solution? I think, I cannot just use Pkg.add("DifferentialEquations") during tests, can I?

ChrisRackauckas commented:
This repo has OrdinaryDiffEq.jl as a test dependency (see test/REQUIRE), and I would recommend just using that directly instead of using DifferentialEquations.jl for testing.

ChrisRackauckas commented:
>Here are my thoughts upon possible solutions for generalising the tools:

Sounds good. I think there should be a fallback too for simply defining the potential and using autodiff on it. We have code already for that fallback, so it would need to be mashed around into this framework. Of course, due to performance concerns you'd have to avoid that if you can though. Also, as ModelingToolkit.jl matures we may want to think about doing some analytical calculations through that, but that's no for 6 months or so down the line.

Mikhail-Vaganov commented:
> This repo has OrdinaryDiffEq.jl as a test dependency (see test/REQUIRE), and I would recommend just using that directly instead of using DifferentialEquations.jl for testing.

@ChrisRackauckas, I got stuck with the test. If I understand it correctly, in order to solve an `ODEProblem`, I need `DifferentialEquations.jl`. I cannot do it just with `OrdinaryDiffEq.jl`. Is it correct?

Also, test/REQUIRE contains only StaticArrays. Do the test tools use the main REQUIRE file of this project first and after that the file from `test` directory?

ChrisRackauckas commented:
All of the types for the dispatch system are defined in DiffEqBase.jl. All of the solvers re-export that library, so things like OrdinaryDiffEq.jl work on their own. It's only default algorithm handling that DifferentialEquations.jl is necessary for. So the problem is here:

https://github.com/JuliaDiffEq/DiffEqPhysics.jl/pull/26/files#diff-55447411abf6b29699c2d37d6ff5eff9R15

that needs an algorithm choice. Just put something standard like `Tsit5()` in there.

Mikhail-Vaganov committed:
use OrdinaryDiffEq instead of DifferentialEquations

Mikhail-Vaganov committed:
moved the animation functions in a separate module

Mikhail-Vaganov committed:
updated readme to new syntax

Mikhail-Vaganov commented:
At last, the tests have become working. 

I have added a recipe for plotting solution of the `NBodyGravProblem`, but it is similar to `orbitplot` function, so I have done it in order to learn how to use the recipes approach. The other animation methods were moved to a separate module so that `NBodyGravitational` does not depend on `Plots`. Nevertheless, animations require tools from `Plot` module - I left it in this pull request simply to show my code.

Mikhail-Vaganov commented:
@ChrisRackauckas, when I and David Sanders had a conversation, I understood, that making tools for the gravitational problem would be quite enough for my first pull request. Now, after these several weeks of working with the task, I understand, that it would be better to have more generic tools in the PR, indeed. 

So what will be the fate of this PR? Should I add or change something so that you could accept my code, or should we just leave it here till the possible GSoC? On the one hand, it is not a good practice to leave a PR unfinished, on the other hand creating generic tools requires much more proper work.

ChrisRackauckas commented:
I don't like having PRs sit. I'll review and likely merge.

ChrisRackauckas commented on nbody_gravitational/GravitationalAnimation.jl:
```diff
@@ -0,0 +1,73 @@
+module GravitationalAnimation
```
this is fine for now but can be done by making an iterator to avoid the dep

ChrisRackauckas commented on nbody_gravitational/test.jl:
```diff
@@ -0,0 +1,41 @@
+end
+
+#the same NBodyGravProblem converted into SecondOrderODEProblem
+solution_simo_3_2nd = solve(problem, Tsit5(), transform_order=2);
```
As a stronger test, use a second-order only method like `DPRKN6` or a symplectic integrator (requires dt)

ChrisRackauckas commented:
You should flip the axis to make the arrays contiguous. Notice you're doing:

```julia
rj = @SVector [rs[j, 1], rs[j, 2], rs[j, 3]]
```

that's using `N` separated values, instead of `rs[1,j]`, `rs[2,j]`, and `rs[3,j]` which are contiguous under column-ordering. This line

```julia
dv[i, 1:3] .= gravitational_acceleration(x.G, u[:, 1:3], m, i, n);
```

is then also not looping contiguously. Nor are any of the setup lines.

ChrisRackauckas commented:
That's probably it. Switch that test algorithm out in the second order case and fix the ordering to be column-order and it'll be ready to merge. And open an issue on handling the animation via iterators.

Mikhail-Vaganov committed:
use the column-major form while handling arrays

ChrisRackauckas commented:
thanks!
